### DISCUSSION TITLE
How to override flow (return from function)?

### DISCUSSION THREAD

**Discussion by saruman9** [Vector35/binaryninja-api#2977] [Q&A] [answered]
I try Binary Ninja demo for my test cases and found that BN don't respect `blx lr` instruction (ARM v7):

![image](https://user-images.githubusercontent.com/4244396/154652413-ddf6357d-0c9c-4811-a3ca-59d8a3c5d780.png)
> Fig. 1. Function don't return after `blx lr`

How can I set override instruction flow to return or set end of the function?

Version: `Binary Ninja 3.0.3233 demo`

**Comment by lwerdna** [comment]
Can you share the binary?

**Reply by saruman9** [reply]
[test.zip](https://github.com/Vector35/binaryninja-api/files/8106533/test.zip)

1. Uncheck `analysis.linearSweep.autorun` parameter.
2. Set `loader.architecture` to ARMv7 (load as mapped).
3. Create a function at 0x0 offset.
4. Go to 0x00000458 offset.

**Comment by saruman9** [comment]
Doesn't answer for a long time mean it's not possible (overriding control flow)?

**Reply by saruman9** [reply]
Firmware or ARM binaries out of scope or has my question already been answered somewhere or are you try to reproduce? Because I don't understand why my question is ignored.

**Reply by lwerdna** [reply]
Hi @saruman9, sorry for the delay, I'm looking into it and will answer soon.

**Comment by lwerdna** [comment, marked as answer]
Analysis doesn't end the function because `blx lr` is considered a call, despite lr being used as the destination operand. The instruction `bx lr` is considered a return, and you can test it by assembling this instruction at 0x458 or patching in bytes "\x1e\xff\x2f\xe1".

**Reply by saruman9** [reply]
I understand that, but the next instruction (`0x0000045c`) never will be executed in the context of returning from `lr` register, because `lr` register will be immediately overwritten (by `0x000003cc`: `bl      sub_4b0`). IDA Pro and Ghidra understand this.

![image](https://user-images.githubusercontent.com/4244396/157717352-da9010ea-3c72-4960-ac3a-110d845cbf99.png)
> Fig. 1. IDA Pro

![image](https://user-images.githubusercontent.com/4244396/157717435-36f2e735-169e-4e85-9e38-e6c6d82b8b2a.png)
> Fig. 2. Ghidra

So if Binary Ninja can't analyze it, can I set end of the function after `blx lr` instruction manually?

**Reply by lwerdna** [reply]
If the link register is being overwritten immediately after `sub_3fc()` returns, I wonder why the compiler generated `blx lr` instead of `bx lr`.

If we make `blx` behave like a return when its operand is `lr` then we fall to the simple obfuscations like:

```
                ; save lr
                ; ...
mov lr, subroutine
blx lr          ; call subroutine
                ; <---- analysis would stop here
                ; ...
                ; restore lr
blx lr          ; return to caller
```

Unfortunately there is no manual way set a function end. Binja thinks of functions as graphs of basic blocks which can be located at arbitrary addresses, so a definitive function end doesn't completely make sense.

Note that the code immediately after the `blx lr` starting at 0x45c is a thumb function `sub_45c()` called from `sub_0()`. So it looks like nonsense when reading it after falling through from ARM `sub_3fc()` but not from thumb `sub_45c()`.

Patching all `blx lr` to `bx lr` is an option. Another is to modify how Binja sees `blx lr` with an architecture hook. There's [an example](https://github.com/Vector35/binaryninja-api/blob/dev/python/examples/arch_hook.py) in the API.

Was this generated by a compiler? If so, and if this is more common than I think, then that will weigh in favor of lifting `blx lr` as a return.

2022-03-11 EDIT: Remove link to architecture hook plugin in private repo, replace with public one from API

**Reply by saruman9** [reply]
> I wonder why the compiler generated blx lr instead of bx lr.
> ...
> Was this generated by a compiler? If so, and if this is more common than I think, then that will weigh in favor of lifting blx lr as a return.

This is not the first time I've encountered this. It is firmware (part of a ROM memory) and a compiler (maybe gcc) is used for generating it. So I never seen that instructions after `blx lr` go further, but your case is possible in theory.

> See [this example](https://github.com/Vector35/automated-binary-ninja/blob/paforvr/exercises/bootrom/answer/bootrom.py#L90-L95) with lines highlighted for "weird call".

I don't have permissions for seeing.

Thank you for detailed answer!

**Reply by lwerdna** [reply]
Hey, sorry about the link to private repo. Here are some public examples of architecture hooks:
* The [api example](https://github.com/Vector35/binaryninja-api/blob/dev/python/examples/arch_hook.py) overrides disassembly to turn "retn" to ret".
* [This example](https://github.com/lwerdna/workbench/tree/master/binja-call-ret-deobf) turns push+ret in x86 into jumps

Here's one for your situation:

```python
from binaryninja.architecture import Architecture, ArchitectureHook
from binaryninja.function import InstructionInfo, InstructionTextToken
from binaryninja.enums import BranchType, InstructionTextTokenType

encoding = b'\x3e\xff\x2f\xe1' # blx lr

class BlxLrOverride(ArchitectureHook):
    def get_instruction_info(self, data, addr):
        if not data.startswith(encoding):
            return super(BlxLrOverride, self).get_instruction_info(data, addr)

        result = InstructionInfo()
        result.length = 4
        result.add_branch(BranchType.FunctionReturn)
        return result

    def get_instruction_low_level_il(self, data, addr, il):
        if not data.startswith(encoding):
            return super(BlxLrOverride, self).get_instruction_low_level_il(data, addr, il)
        
        il.append(il.ret(il.reg(4, 'lr')))
        return 4

BlxLrOverride(Architecture['armv7']).register()
```

**Comment by lwerdna** [comment]
Thank you for reporting the issue. We've decided to lift `blx lr` as return: https://github.com/Vector35/arch-armv7/commit/662d1009d707ae7b9010ae1e57cf6829d87b12ff

![919983857](https://user-images.githubusercontent.com/474191/157757798-16c94b7b-c461-44f2-b6ce-d7021502f078.png)

If you have a normal binaryninja license, please update to latest DEV release.

Demo versions have to wait until STABLE release.

2022-03-11 EDIT: Change commit link to public arch-armv7 repo.

**Reply by saruman9** [reply]
Thank you very much!

**Reply by saruman9** [reply]
After fixing the link I tested the plugin. It works well. I think, that Thumb's `blx lr` instruction needed to lift as return too ([1](https://github.com/Vector35/arch-armv7/blob/662d1009d707ae7b9010ae1e57cf6829d87b12ff/thumb2_disasm/arch_thumb2.cpp#L225), [2](https://github.com/Vector35/arch-armv7/blob/662d1009d707ae7b9010ae1e57cf6829d87b12ff/thumb2_disasm/il_thumb2.cpp#L601)). 

![image](https://user-images.githubusercontent.com/4244396/157935402-47e9fd2f-d9a5-4b67-a25f-04ff2d61e785.png)
> Fig. 1. `blx lr` should return

**Comment by bpotchik** [comment]
As of `3.6.4594-dev`, we have added the ability to interpret 'blx lr' given the context of the target register. It should now be translated to a call, or a function return in the correct scenarios.
